glib: fix tests

alyssais opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

alyssais committed:
glib: remove unnecessary substituteInPlace

/bin/true hasn't been hardcoded in this file since Glib 2.55.0.
2.56.0, the first stable version to include this change, was released
in 2018.

Fixes: 0973618e4ec ("gnome3: automated update")

alyssais committed:
glib: fix paths of tests in sed

Since the switch to Meson, we're already in the build directory in
preCheck, so we need to update the relative paths we're using
accordingly.

Fixes: 0149459f424 ("glib: 2.56.0 → 2.58.1")

alyssais committed:
glib: fix tests by patching Python shebangs

This has to happen in postConfigure rather than postPatch, because the
files are generated by the configure phase.

alyssais committed:
glib: move test substitutions into postPatch

Doing this in preCheck means that the tests will already have been
built before the substitutions are made, so they'll have to be rebuilt
after the substitutions.  Move the substitution earlier in the build,
to avoid this unnecessary extra work.

I've made the substitutions run even if tests aren't being run, to
avoid a repeat of the situation where a glib update breaks the tests
due to paths changing.

alyssais committed:
glib: skip g-file-info test if atime unsupported

atime=off is probably not uncommon among Nix users using ZFS.

alyssais commented:
@OfBorg build glib.override{doCheck=true;}

risicle commented:
Could use the `finalAttrs.finalPackage` trick to add a `passthru.tests.withChecks` which would allow you to call it more easily (and perhaps make it more likely for people to try it)

alyssais commented:
> Could use the `finalAttrs.finalPackage` trick to add a `passthru.tests.withChecks` which would allow you to call it more easily (and perhaps make it more likely for people to try it)

How would I do that?  I tried

```
    tests.withChecks = finalAttrs.finalPackage.override { doCheck = true; };
```

But this doesn't work because `finalPackage` is not overridable.

SuperSandro2000 commented on pkgs/development/libraries/glib/default.nix:
```diff
@@ -154,11 +161,25 @@ stdenv.mkDerivation rec {
     patchShebangs glib/gen-unicode-tables.pl
     patchShebangs glib/tests/gen-casefold-txt.py
     patchShebangs glib/tests/gen-casemap-txt.py
```
```suggestion
    patchShebangs glib/gen-unicode-tables.pl \
      glib/tests/gen-casefold-txt.py \
      glib/tests/gen-casemap-txt.py
```
Unrelated but we can simplify this

alyssais replied:
This PR is big enough as it is, so I don't want to start adding in unrelated changes.

SuperSandro2000 commented on pkgs/development/libraries/glib/default.nix:
```diff
@@ -154,11 +161,25 @@ stdenv.mkDerivation rec {
       --replace "libintl, " ""
   '';
 
+  postConfigure = ''
+    patchShebangs gio/gdbus-2.0/codegen/gdbus-codegen gobject/glib-{genmarshal,mkenums}
+  '';
```
This should be done in postPatch

alyssais replied:
Please read the commit messages.

risicle commented:
> But this doesn't work because `finalPackage` is not overridable.

Oh yes, can't use `.override`, but you should be able to use `.overrideAttrs`.

7c6f434c reviewed (approved):
Diff looks good

alyssais committed:
glib: unconditionally apply test patch

This patch shouldn't have any effect on builds that don't run the
tests, so it doesn't need to be conditionally applied.  Applying it
unconditionally will make it easier to notice if it breaks e.g. when
updating glib.

alyssais committed:
glib: use finalAttrs instead of rec

This will be important when we replace the doCheck option with an
overrideAttrs, because then we can use it in passthru.tests.

alyssais committed:
glib.tests.withChecks: init

This will allow OfBorg to run the GLib tests, which it didn't do
before.  To do this correctly for the glib we're building, we need to
override it rather than taking glib as an input.
finalAttrs.finalPackage is not overrideable, so instead we need to use
overrideAttrs.  Since we can now consult finalAttrs.doCheck, there's
no longer any need for a separate doCheck option, so I've removed it
in favour of overrideAttrs.

SuperSandro2000 commented:
Some tests seem to only expected fail on x86_64-linux, not any other platform. Maybe ignore those on other platforms?

alyssais commented:
> Some tests seem to only expected fail on x86_64-linux, not any other platform. Maybe ignore those on other platforms?

I'm not sure what you mean, but my goal here is just to make it possible to run the tests at all, not to make them all pass.

jtojnar commented:
Changes look good overall but it might be better to check for `ST_NOATIME` as suggested upstream: https://gitlab.gnome.org/GNOME/glib/-/merge_requests/2866#note_1544420

SuperSandro2000 commented:
> I'm not sure what you mean, but my goal here is just to make it possible to run the tests at all, not to make them all pass.

They don't need to pass all but we should disable the 3 tests in total that fail on other platforms.

jtojnar reviewed (approved):
Sounds good to me with explanation from https://gitlab.gnome.org/GNOME/glib/-/merge_requests/2866#note_1544862

alyssais commented:
> They don't need to pass all but we should disable the 3 tests in total that fail on other platforms.

We shouldn't be disabling tests just because they fail.  We should be investigating them, and then either disabling them if they're Nix-specific, or reporting them upstream if they are not.  As these failures are for architectures I don't use, I can't do that, but that's fine, because these tests are not run by default at the moment anyway.  (For test failures that happen on hardware I do have access to, I [have done the upstream work](https://gitlab.gnome.org/GNOME/glib/-/merge_requests/2866).)
